icefoganalytics/travel-authorization
# Add Admin Interface To View a Travel Authorization Request – With Ability To Approve/Decline the Request

# Context

Add admin interface to view a travel authorization request from another user, and approve/decline that request.

Notes:
- http://localhost:8080/travel-request exists, and is a global table of all travel requests you can see. It is not related to the feature being built here, at this time anyway.

## Out of Scope

- Ability to edit the travel authorization request of another user.
- Ability to assign the request to another user.

## Proposed Implementation

- Add a /travel-requests/:travelAuthorizationId/details front-end route, with a read only view of the a user's travel authorization.
- This view should have call to action at the bottom with Approve/Decline actions.
- This should probably be very similar, or identical to the `/my-travel-requests/<id>/details` view, but not bound to the current user. I might be able to abstract some of the components so they can be used in both places.

# Implementation

Add Approve/Deny Controllers
Add TravelAuthorizationActionLog model to track approval/denial actors.
Upgrade travel authorization store pattern. You can now access a "current" travel authorization.
Fix bug in pre-approved-travel-requests readonly modeling.
Fix reactivity bug between trip types and summary panel. Vue2 has poor deep reactivity support so you must replace the entire stops list when you edit a stop, or reactivity won't propagate. In theory "splice" will work too, but it didn't trigger reactivity into the Vue Devtools plugin, so I just used replace everywhere for simplicity.

# Screenshots

![image](https://github.com/icefoganalytics/travel-authorization/assets/23045206/896e0948-7067-4bf3-b571-0758e0d6e6bd)


# Testing Instructions

1. Boot the app via `dev up`
2. Log in to the app at http://localhost:8080.
3. Go to the "My Travel Requests" page via the top nav.
4. Create a new Travel Authorization via the "+ Travel Authorization" button.
5. Fill in the Purpose section, and check the "Purpose" syncs with the top header panel.
6. Check that the "Final Destination" syncs with the top header panel.
7. Check that the "Travel Method" defaults to `Aircraft`.
8. Check that the "Type of Accommodation" defaults to `Hotel` for the first stop, and blank for the second stop.
9. Save the draft.
10. After saving, generate an estimate (don't generate estimate first, or it'll wipe all your data see https://github.com/icefoganalytics/travel-authorization/issues/92)
11. After generating the estimate, submit the travel authorization to yourself for approval. This seems like something we would need to block? But it does make testing easier.
12. After you submit the authorization, go to the "Manager View" from the top drop down nav.
13. You will see your travel authorization request in the "Pending Approvals" widget in the top left.
14. Click on the travel authorization to access the admin view of that request. We may want to add an "action" column with a button that you can click on instead, as it's much clearer user experience.
15. Check that you get a nice read only view of the travel authorization you submitted, with an extra panel called "Management". The UI is a little light here, but there are a bunch more things that'll get added.
16. Click the Approve button to approve the travel request. We might want to add a "status" chip the page by the user chip at the top?
17. Click the "Back" button to go back to the manager view.
18. Your travel authorization will now be in the bottom right widget under "Approved Trips".
19. Repeat the process, but this time "Deny" the travel request.
20. Click the "Back" button to go back to the manager view.
21. The travel authorization will be gone from the "Pending Approvals" widget in the top left. There isn't currently a "Denied Trips" panel, but it would be _very_ easy to add one.
